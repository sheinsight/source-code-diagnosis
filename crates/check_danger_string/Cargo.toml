[package]
name = "check_danger_string"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
napi-derive = { workspace = true }
napi = { workspace = true }
serde = { workspace = true }
oxc_ast = { workspace = true }
utils = { path = "../utils" }
beans = { path = "../beans" }
anyhow = { workspace = true }
parking_lot = { workspace = true }
oxc_span = {workspace = true}
pathdiff = { workspace = true }
camino = { workspace = true }
wax = { workspace = true }
rayon = { workspace = true }
oxc_allocator = { workspace = true }
oxc_parser = { workspace = true }
oxc_semantic = { workspace = true }
insta = { workspace = true }
