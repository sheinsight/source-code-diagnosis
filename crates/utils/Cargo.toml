[package]
name = "utils"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
anyhow = "1.0.88"
napi-derive = { workspace = true }
napi = { workspace = true }
rayon = "1.10.0" 
wax = "0.6.0"
ropey = "1.0.0"
pathdiff = {workspace = true}
beans = { path = "../beans" }
log = { workspace = true }
camino = { workspace = true }

oxc = {workspace = true}