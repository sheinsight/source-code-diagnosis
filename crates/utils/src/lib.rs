mod glob;
mod semantic_builder;

use std::{
  fs,
  io::{BufReader, Read},
};

pub use glob::{glob_by, GlobArgs, GlobJsArgs};
pub use semantic_builder::{
  source_type_from_path, SemanticBuilder, SemanticHandler,
};

pub fn read_file_content(path: &std::path::Path) -> anyhow::Result<String> {
  let file = fs::File::open(path)?;
  let mut reader = BufReader::with_capacity(1024 * 1024, file); // 1MB buffer
  let mut content = String::new();
  reader.read_to_string(&mut content)?;
  Ok(content)
}

pub fn is_ts_video(path: &std::path::Path) -> bool {
  if let Ok(mut file) = fs::File::open(path) {
    let mut buffer = [0; 4];
    if file.read_exact(&mut buffer).is_ok() {
      // TS 视频文件的魔数是 0x47
      return buffer[0] == 0x47;
    }
  }
  false
}
