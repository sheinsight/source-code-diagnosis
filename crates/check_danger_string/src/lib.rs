use anyhow::Result;
use beans::AstNode;
use napi_derive::napi;
use oxc_allocator::Allocator;
use oxc_ast::AstKind;
use oxc_parser::Parser;
use oxc_semantic::SemanticBuilder;
use serde::Serialize;
use utils::{glob_by, source_type_from_path, GlobArgs};

#[napi(object)]
#[derive(Debug, Serialize)]
pub struct CheckDangerResponse {
  pub file_path: String,
  pub items: Vec<CheckDangerResponseItem>,
  pub errors: Vec<String>,
}

#[napi(object)]
#[derive(Debug, Serialize)]
pub struct CheckDangerResponseItem {
  pub raw_value: String,
  pub match_danger_string: String,
  pub ast_node: AstNode,
}

pub fn check_danger_strings(
  danger_strings: Vec<String>,
  args: GlobArgs,
) -> Result<Vec<CheckDangerResponse>> {
  let responses = glob_by(
    |path| {
      let relative_path = pathdiff::diff_paths(path, &args.cwd)?;

      let relative_path_str =
        utils::win_path_to_unix(relative_path.display().to_string().as_str());

      let source_code = match std::fs::read_to_string(path) {
        Ok(content) => content,
        Err(err) => {
          return Some(CheckDangerResponse {
            file_path: relative_path_str,
            items: vec![],
            errors: vec![format!("文件读取错误: {}", err)],
          });
        }
      };

      let allocator = Allocator::default();
      let source_type = source_type_from_path(path);

      let parser = Parser::new(&allocator, &source_code, source_type);
      let parse = parser.parse();

      if !parse.errors.is_empty() {
        return Some(CheckDangerResponse {
          file_path: relative_path_str,
          items: vec![],
          errors: vec![format!("解析错误: {:?}", parse.errors)],
        });
      }

      let program = allocator.alloc(parse.program);

      let semantic_return = SemanticBuilder::new()
        .with_check_syntax_error(false)
        .build(program);

      let semantic = semantic_return.semantic;

      let responses = semantic
        .nodes()
        .into_iter()
        .filter_map(|node| {
          if let AstKind::StringLiteral(string_literal) = node.kind() {
            let value = string_literal.value.to_string();

            let ast_node = AstNode::with_source_and_span(
              &semantic.source_text(),
              &string_literal.span,
            );

            for danger_string in danger_strings.iter() {
              if value.contains(danger_string) {
                return Some(CheckDangerResponseItem {
                  raw_value: value.clone(),
                  match_danger_string: danger_string.to_string(),
                  ast_node,
                });
              }
            }
          }
          None
        })
        .collect::<Vec<_>>();

      Some(CheckDangerResponse {
        file_path: relative_path_str,
        items: responses,
        errors: vec![],
      })
    },
    &args,
  )?;

  Ok(responses)
}
