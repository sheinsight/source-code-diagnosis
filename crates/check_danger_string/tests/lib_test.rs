use check_danger_string::check_danger_strings;
use utils::GlobJsArgs;

#[test]
fn test_check_danger_string() -> anyhow::Result<()> {
  let result = check_danger_strings(
    vec!["taobao.com".to_string()],
    GlobJsArgs {
      cwd: "tests/fixtures/danger".to_string(),
      pattern: None,
      ignore: None,
    }
    .into(),
  )?;

  insta::assert_debug_snapshot!(result);

  Ok(())
}
