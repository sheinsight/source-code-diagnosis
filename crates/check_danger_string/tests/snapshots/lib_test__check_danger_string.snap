---
source: crates/check_danger_string/tests/lib_test.rs
expression: result
snapshot_kind: text
---
[
    CheckDangerResponse {
        file_path: "index.jsx",
        items: [
            CheckDangerResponseItem {
                raw_value: "taobao.com",
                match_danger_string: "taobao.com",
                ast_node: AstNode {
                    span: Span {
                        start: 0,
                        end: 12,
                    },
                    loc: Location {
                        start: Position {
                            line: 1,
                            col: 1,
                        },
                        end: Position {
                            line: 1,
                            col: 13,
                        },
                    },
                },
            },
        ],
        errors: [],
    },
]
